src/Substructure.jl

```diff
@@ -0,0 +1,304 @@
+    recluster_method::JetAlgorithm.Algorithm
+end
+
+"""
+    mass_drop(jet, clusterseq, tag) -> PseudoJet
+
+Identifies subjets in a jet that pass the mass drop tagging condition.
+The method stops at the first jet satisfying the mass and distance thresholds.
+
+Arguments:
+- jet: PseudoJet instance representing the jet to tag
+- clusterseq: ClusterSequence with jet clustering history
+- tag: MassDropTagger instance providing mass drop parameters
+
+Returns:
+- PseudoJet: The jet (or subjet) satisfying the mass drop conditions, if tagging is successful, otherwise a zero-momentum PseudoJet
+"""
+
+function mass_drop(jet::PseudoJet, clusterseq::ClusterSequence, tag::MassDropTagger)
+    all_jets = clusterseq.jets
+    hist = clusterseq.history
+
+    while true
+        had_parents, p1, p2 = has_parents(jet, clusterseq)
+
+        if had_parents
+            parent1 = all_jets[hist[p1].jetp_index]
+            parent2 = all_jets[hist[p2].jetp_index]
+
+            if m2(parent1) < m2(parent2)
+                p1, p2 = p2, p1
+                parent1, parent2 = parent2, parent1
+            end
+
+            if m2(parent1) < m2(jet) * tag.mu^2 &&
+               kt_distance(parent1, parent2) > tag.y * m2(jet)
+                return jet
+            else
+                jet = parent1
+            end
+
+        else
+            return PseudoJet(0.0, 0.0, 0.0, 0.0)
+        end
+    end
+end
+
+"""
+    soft_drop(jet, clusterseq, tag) -> PseudoJet
+
+Applies soft-drop grooming to remove soft, wide-angle radiation from jets.
+This function reclusters the jet and iteratively checks the soft-drop condition on subjets.
+
+Arguments:
+- jet: PseudoJet instance to groom
+- clusterseq: ClusterSequence containing jet history
+- tag: SoftDropTagger instance with soft-drop parameters
+
+Returns:
+- PseudoJet: Groomed jet or zero-momentum PseudoJet if grooming fails
+"""
+
+function soft_drop(jet::PseudoJet, clusterseq::ClusterSequence,
+                   tag::SoftDropTagger)
+    rad = tag.cluster_rad
+    new_clusterseq = recluster(jet, clusterseq; R = rad, algorithm = JetAlgorithm.CA)
+    new_jet = sort_jets!(inclusive_jets(new_clusterseq; T = PseudoJet))[1]
+
+    all_jets = new_clusterseq.jets
+    hist = new_clusterseq.history
+
+    while true
+        had_parents, p1, p2 = has_parents(new_jet, new_clusterseq)
+
+        if had_parents
+            parent1 = all_jets[hist[p1].jetp_index]
+            parent2 = all_jets[hist[p2].jetp_index]
+
+            if m2(parent1) < m2(parent2)
+                p1, p2 = p2, p1
+                parent1, parent2 = parent2, parent1
+            end
+
+            pti = pt2(parent1)^0.5
+            ptj = pt2(parent2)^0.5
+
+            if min(pti, ptj) / (pti + ptj) >
+               tag.zcut * (deltaR(parent1, parent2) / rad)^tag.b
+                return new_jet
+            else
+                new_jet = parent1
+            end
+
+        else
+            return PseudoJet(0.0, 0.0, 0.0, 0.0)
```

Comment:
I am not completely convinced by this return type - it might be better in this case to return `nothing`. But we need to check this is ok from a performance point of view. Could you check that?